feat(Controls): Adds a way to add more options to the `ctrls` element.

AlexanderGeere opened the pull request:
## Description
This PR introduces a `ui.elements.controls` element as well as  a new utility called `controls`. 

The default view now calls `mapp.ui.elements.controls` to create the control tabs and the panel associated to the tab. The default controls are `layer` and `locations`, If no controls are specified in the configuration.   

custom controls can also be created by specifying a plugin and a control as follows:

```json
"locale": {
   "plugins": ["${LOCAL}/bugs_testing/test/my_control.js"]
   "controls": ["layers",
                  {"key": "my_control", "icon": "history"}, 
                  {"key": "locations", "icon": "location_on", "style": {"display": "block"}}]
   "my_control": {}
}
```

and then in the plugin.

```javascript
mapp.controls.my_control = () => {console.log('This is a custom control')}
```

This would yield a `ctrls` element that looks as follows:
<img width="343" height="41" alt="screenshot-2025-11-05_15-41-46" src="https://github.com/user-attachments/assets/b660f2d4-b0bb-4da8-846d-fec4734f567a" />



## GitHub Issue
#2450 

## Type of Change
- ✅ New feature (non-breaking change which adds functionality)
- ✅ Documentation

## How have you tested this?
This can be tested on `bugs_testing/dataviews/dataviews.json`

## Testing Checklist
- ✅ Existing Tests still pass
- ✅ Ran locally on my machine

## Code Quality Checklist
- ✅ My code follows the guidelines of XYZ
- ✅ My code has been commented
- ✅ Documentation has been updated
- ✅ New and existing unit tests pass locally with my changes
- ✅ Main has been merged into this PR

dbauszus-glx reviewed (changes requested):
There shouldn't be any breadcrumbing by creating a new control.gazetteer module/method in the library and then changing the ui gazetteer to lead to the new module.

Any ui stuff should be in the ui module. The mapp.ui module does depend on the mapp library. But the mapp library must not have any dependencies on the ui module.

The idea is that the mapp library can be used to build an interface with your own custom elements without loading the ui extension of the mapp library.

We should aim to reduce inline styling not add more as in the listview module.

The css for the control tabs element should be defined in the ui.css sheets as much as possible.

dbauszus-glx commented:
This is way too complicated at the moment.

The control panel element should not require nested tabs. The tabs should be generated when required.

```
    <div id="ctrls">
      <div id="ctrl-tabs" class="hover"></div>
      <div id="ctrl-panel"></div>
    </div>
```

Decorating the mapview with controlTabs and controlPanels elements, then assigning mapview.locale.controls, then running mapp.ui.elements.controls which doesn't return anything, then defining the layersTab, then do other things before at the end loading controls from the mapp library into the ui.controls element feels too many steps.

dbauszus-glx commented:
I put this into draft for now and removed the mapp.control logic. The control element should be handled by ui.elements.control.

This is a decorator method which returns a decorated object.

The decorator adds an add method to add controls to the object.

```js
    //Create tabs and panels for controls.
    const control = mapp.ui.elements.control({
      target: document.getElementById('ctrls')
    })

    const layersPanel = mapp.utils.html.node`<div id="layers" class="active">`

    control.add({
      key: 'layers',
      panel: layersPanel
    })
    
    mapp.ui.layers.listview({
      target: layersPanel,
      layers: mapview.layers,
    });

    const locationsPanel = mapp.utils.html.node`<div id="locations">`

    control.add({
      key: 'locations',
      icon: 'location_on',
      panel: locationsPanel
    })
    
    mapp.ui.locations.listview({
      target: locationsPanel,
      mapview,
    });   
```

The ui.gazetteer method needs to be reviewed proper, the params/properties were all over the shop.

There is no need for a mapp.controls.listview method which is dependent on the ui.listview methods.

I added an elements/css module and removed the relevant css for this element from the view document head.

AlexanderGeere commented:
@dbauszus-glx 

I added the gazetteer by using the `control.add` and made it so the location tab is always there. 
The gazetteer doesnt replace the lcoation tab anymore.

I made a class for the no location selection text i added. 

I added the control element to the mapview in the, I thought we could just call add in the plugins or elsewhere if we wanted to add more controls down the line. 

I will do the documentation if you are happy with the structure.

simon-leech reviewed (dismissed):
<img width="295" height="63" alt="image" src="https://github.com/user-attachments/assets/4675bf1a-0c40-40e7-affc-688e01a0cf0d" />

The hover text should be more clear than 'gazetteer' . Maybe something like 'Search for a location' ? 

The rest is working fine for me though! 
Added to a custom view a `controlPanel` and was nice and easy, added a custom control 👍 
Gazetteer still working fine in a custom view too.
